independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java

```diff
@@ -355,46 +333,61 @@ public URL getResource(String unsanitisedName) {
+        ClassPathResourceIndex classPathResourceIndex = getClassPathResourceIndex();
+        if (classPathResourceIndex.isBanned(name)) {
             return null;
         }
         //TODO: because of dev mode we iterate, to see if any resources were added
         //not for .class files though, adding them causes a restart
         //this is very important for bytebuddy performance
         boolean endsWithTrailingSlash = unsanitisedName.endsWith("/");
         if (name.endsWith(".class") && !endsWithTrailingSlash) {
-            ClassPathElement[] providers = state.loadableResources.get(name);
-            if (providers != null) {
-                final ClassPathResource resource = providers[0].getResource(name);
+            ClassPathElement classPathElement = classPathResourceIndex.getFirstClassPathElement(name);
+            if (classPathElement != null) {
+                final ClassPathResource resource = classPathElement.getResource(name);
                 if (resource == null) {
-                    throw new IllegalStateException(providers[0] + " from " + getName() + " (closed=" + this.isClosed()
-                            + ") was expected to provide " + name + " but failed");
+                    throw new IllegalStateException(
+                            classPathElement + " from " + getName() + " (closed=" + this.isClosed()
+                                    + ") was expected to provide " + name + " but failed");
                 }
                 return resource.getUrl();
             }
         } else {
-            for (ClassPathElement i : elements) {
-                ClassPathResource res = i.getResource(name);
-                if (res != null) {
-                    //if the requested name ends with a trailing / we make sure
-                    //that the resource is a directory, and return a URL that ends with a /
-                    //this matches the behaviour of URLClassLoader
-                    if (endsWithTrailingSlash) {
-                        if (res.isDirectory()) {
-                            try {
-                                return new URL(res.getUrl().toString() + "/");
-                            } catch (MalformedURLException e) {
-                                throw new RuntimeException(e);
-                            }
+            URL url = getClassPathElementResourceUrl(normalPriorityElements, name, endsWithTrailingSlash);
+            if (url != null) {
+                return url;
+            }
+            url = getClassPathElementResourceUrl(lesserPriorityElements, name, endsWithTrailingSlash);
+            if (url != null) {
+                return url;
+            }
+        }
+        return parent.getResource(unsanitisedName);
+    }
+
+    private static URL getClassPathElementResourceUrl(List<ClassPathElement> classPathElements, String name,
+            boolean endsWithTrailingSlash) {
+        for (ClassPathElement classPathElement : classPathElements) {
```

Comment:
Can we use a good old `for(i=0...)` here in order to cut down on allocations?

Reply:
Sure we could but I don't think it's worse than it used to be. So ideally, let's do that in a follow-up PR?

Reply:
Sure